homeassistant/components/roon/media_browser.py

```diff
@@ -0,0 +1,165 @@
+"""Support to interface with the Plex API."""
+import logging
+
+from homeassistant.components.media_player import BrowseMedia
+from homeassistant.components.media_player.const import (
+    MEDIA_CLASS_DIRECTORY,
+    MEDIA_CLASS_PLAYLIST,
+    MEDIA_CLASS_TRACK,
+)
+from homeassistant.components.media_player.errors import BrowseError
+
+
+class UnknownMediaType(BrowseError):
+    """Unknown media type."""
+
+
+EXCLUDE_ITEMS = {
+    "Play Album",
+    "Play Artist",
+    "Play Playlist",
+    "Play Composer",
+    "Play Now",
+    "Play From Here",
+    "Queue",
+    "Start Radio",
+    "Add Next",
+    "Play Radio",
+    "Play Work",
+    "Settings",
+    "Search",
+    "Search Tidal",
+    "Search Qobuz",
+}
+
+# Maximum number of items to pull back from the API
+ITEM_LIMIT = 3000
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def browse_media(zone_id, roon_server, media_content_type=None, media_content_id=None):
+    """Implement the websocket media browsing helper."""
+    try:
+        _LOGGER.debug("browse_media: %s: %s", media_content_type, media_content_id)
+        if media_content_type in [None, "library"]:
+            return library_payload(roon_server, zone_id, media_content_id)
+
+    except UnknownMediaType as err:
+        raise BrowseError(
+            f"Media not found: {media_content_type} / {media_content_id}"
+        ) from err
+
+
+def item_payload(roon_server, item, list_image_id):
+    """Create response payload for a single media item."""
+
+    title = item.get("title")
+    subtitle = item.get("subtitle")
+    if subtitle is None:
+        display_title = title
+    else:
+        display_title = f"{title} ({subtitle})"
+
+    image_id = item.get("image_key") or list_image_id
+
+    image = None
+    if image_id:
+        image = roon_server.roonapi.get_image(image_id)
+
+    media_content_id = item.get("item_key")
+    media_content_type = "library"
+
+    hint = item.get("hint")
+    if hint == "list":
+        media_class = MEDIA_CLASS_DIRECTORY
+        can_expand = True
+    elif hint == "action_list":
+        media_class = MEDIA_CLASS_PLAYLIST
+        can_expand = False
+    elif hint == "action":
+        media_content_type = "track"
+        media_class = MEDIA_CLASS_TRACK
+        can_expand = False
+    else:
+        # Roon API says to treat unknown as a list
+        media_class = MEDIA_CLASS_DIRECTORY
+        can_expand = True
+        _LOGGER.warning("Unknown hint %s - %s", title, hint)
+
+    if title in EXCLUDE_ITEMS:
+        return None
```

Comment:
Move this check to the beginning to avoid doing unnecessary work. 

Feels a little weird that a payload generating function does this check. Shouldn't we call this function to begin with if it's excluded?

Reply:
I've done as you suggested and moved it into the calling code.